### PR TITLE
Add qt.qmlplugin rule and support of qmltypesregistrar

### PR DESCRIPTION
**Pull request by Arthapz**

### CODE REVIEW

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        if not os.isexec(qmltyperegistrar) and qt.libexecdir then
+            qmltyperegistrar = path.join(qt.libexecdir, is_host("windows") and "qmltyperegistrar.exe" or "qmltyperegistrar")
+        end
+        if not os.isexec(qmltyperegistrar) and qt.libexecdir_host then
+            qmltyperegistrar = path.join(qt.libexecdir_host, is_host("windows") and "qmltyperegistrar.exe" or "qmltyperegistrar")
+        end
+        assert(qmltyperegistrar and os.isexec(qmltyperegistrar), "qmltyperegistrar not found!")
+
+        -- set targetdir
+        local importname = target:values("qml.plugin.importname")
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:40]: You should check if it is nil

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        for _, dir in pairs(importname:split(".", { plain = true })) do
+            targetdir = path.join(targetdir, dir)
+        end
+
+        os.mkdir(targetdir)
+        target:set("targetdir", targetdir)
+
+        -- add qmldir
+        local qmldir = target:values("qml.plugin.qmldir")
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:51]: remove empty line

Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:51]: please rename to `qt.qmlplugin.import_name`

File: xmake/rules/qt/qmltyperegistrar/xmake.lua (hunk 1 above)
Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:40]: please rename to `qt.qmlplugin.import_name`

File: xmake/rules/qt/qmltyperegistrar/xmake.lua (hunk 2 above)
Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:51]: please rename to `qt.qmlplugin.qmldir`

Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:51]: please rename to `qt.qmlplugin.qmldir`

Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:51]: please rename to `qt.qmlplugin.qmldir`

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        end
+        assert(qmltyperegistrar and os.isexec(qmltyperegistrar), "qmltyperegistrar not found!")
+
+        -- set targetdir
+        local importname = target:values("qml.plugin.importname")
+        local targetdir = path.join(target:targetdir(), "imports")
+        for _, dir in pairs(importname:split(".", { plain = true })) do
+            targetdir = path.join(targetdir, dir)
+        end
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:45]: please remove empty line

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        os.mkdir(sourcefile_dir)
+        target:data_set("qml.plugin.sourcefile", sourcefile)
+
+        -- add moc arguments
+        target:add("qt.moc.flags", "--output-json")
+
+        -- save qmltyperegistrar
+        target:data_set("qt.qmltyperegistrar", qmltyperegistrar)
+        -- save qmltyperegistrar
+        target:data_set("qml.plugin.qmltyperegistrar", qmltyperegistrar)
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:67]: please rename to `qt.qmlplugin.qmltyperegistrar`

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+     on_buildcmd_files(function(target, batchcmds, sourcebatch, opt)
+        assert(target:values("qml.plugin.importname"), "QML plugin import name not set")
+
+        -- setup qmltyperegistrar arguments
+        local qmltyperegistrar = target:data("qt.qmltyperegistrar")
+        local sourcefile = target:data("qml.plugin.sourcefile")
+
+        local importname = target:values("qml.plugin.importname")
+        local majorversion = target:values("qml.plugin.majorversion") or 1
+        local minorversion = target:values("qml.plugin.minorversion") or 0
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:79]: please rename to `qt.qmlplugin.xxx`

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        for _, mocedfile in ipairs(sourcebatch.sourcefiles) do
+            target:add("includedirs", path.directory(mocedfile))
+            local basename = path.basename(mocedfile)
+            local filename_moc = "moc_" .. basename .. ".cpp"
+            if mocedfile:endswith(".cpp") then
+                filename_moc = basename .. ".moc"
+            end
+
+            local sourcefile_moc = target:autogenfile(path.join(path.directory(mocedfile), filename_moc))
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:91]: please remove empty line

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        local metatypefiles = {}
+        for _, mocedfile in ipairs(sourcebatch.sourcefiles) do
+            target:add("includedirs", path.directory(mocedfile))
+            local basename = path.basename(mocedfile)
+            local filename_moc = "moc_" .. basename .. ".cpp"
+            if mocedfile:endswith(".cpp") then
+                filename_moc = basename .. ".moc"
+            end
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:89]: please remove empty line

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        batchcmds:show_progress(opt.progress, "${color.build.object}generate.qt.qmltyperegistrar %s", path.filename(sourcefile))
+        qmltype_source = os.vrunv(qmltyperegistrar, table.join(args, metatypefiles))
+
+        -- add objectfile
+        local objectfile = target:objectfile(sourcefile)
+        table.insert(target:objectfiles(), objectfile)
+
+        -- compile sourcefile
+        batchcmds:show_progress(opt.progress, "${color.build.object}compile.qt.qmltyperegistrar %s", path.filename(sourcefile))
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:113]: please remove empty line

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        batchcmds:compile(sourcefile, objectfile)
+
+        batchcmds:add_depfiles(sourcefile)
+        batchcmds:set_depmtime(os.mtime(objectfile))
+        batchcmds:set_depcache(target:dependfile(objectfile))
+    end)
+
+    after_build(function(target)
+        local qmldir = target:values("qml.plugin.qmldir")
+
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:123]: and here

File: tests/projects/qt/quickplugin/src/Foo.cpp [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+Foo::Foo(QObject *parent) : QObject { parent }, m_bar{ 0 } {
+}
+
+int Foo::bar() const noexcept {
+    return m_bar;
+}
+
+void Foo::setBar(int bar) noexcept {
+    if(bar == m_bar) return;
```
Comment (waruqi) [inline tests/projects/qt/quickplugin/src/Foo.cpp:11]: please add space in `if (bar`

File: tests/projects/qt/quickplugin/xmake.lua [hunk 11]
```diff
@@ -0,0 +1,12 @@
+add_rules("mode.debug", "mode.release")
+
+target("demo")
+    add_rules("qt.qmlplugin")
+    add_headerfiles("src/*.h")
+    add_files("src/*.cpp")
+
+    if is_plat("windows") then
+        add_cxxflags("/permissive-")
```
Comment (waruqi) [inline tests/projects/qt/quickplugin/xmake.lua:9]: If qt has to rely on this compile option, then you can set it in qt/env/xmake.lua. If it is only an optional option, then remove it here. Leave it to the user to set it additionally as needed.

File: xmake/rules/qt/qmltyperegistrar/xmake.lua [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        for _, dir in pairs(importname:split(".", { plain = true })) do
+            targetdir = path.join(targetdir, dir)
+        end
+        os.mkdir(targetdir)
+        target:set("targetdir", targetdir)
+
+        -- add qmldir
+        local qmldir = target:values("qt.qmlplugin.qmldir")
+        if qmldir then
+            target:add("installfiles", qmldir)
```
Comment (waruqi) [inline xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: installfiles is directory? I don't remember supporting it.

Comment (Arthapz) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: qmldir is not a directory, this is a file

Comment (Arthapz) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: (and it has to be named qmldir, not my name)

Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: why?

Comment (arkena00) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: qt stuffs

Comment (waruqi) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: I think it will be better of we use `qt.qmlplugin.qmldirfile`

Comment (Arthapz) [reply xmake/rules/qt/qmltyperegistrar/xmake.lua:53]: Okay

**Comment by waruqi** [comment]
Thanks, Can you add a line in changelog.md?

**Comment by Arthapz** [comment]
> 

do i update the changelog and make a new pull request ?

**Comment by waruqi** [comment]
I have added it.

**Pull request merged**